### PR TITLE
HDDS-9041. Intermittent Delete root failed.

### PR DESCRIPTION
**Pull request by devmadhuu**
## What changes were proposed in this pull request?

Intermittent Delete root failed where it is observed that listStatus for FSO returns deleted files. This is causing flaky for some test cases.
Problem: listStatus called before flush of double buffer can return deleted key / directory present in cache. And also old information from db.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9041

## How was this patch tested?

This patch is tested by executing existing flaky test cases which is now running fine.

### CODE REVIEW

**Comment by devmadhuu** [comment]
@sumitagrawl  - Pls review.

**Review by sumitagrawl — APPROVED** [review]
@devmadhuu Thanks for working over this, LGTM

**Review by duongkame — APPROVED** [review]
Thanks for the fix, @devmadhuu . LGTM.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java [hunk 1]
```diff
@@ -214,11 +215,12 @@ public Collection<OzoneFileStatus> listStatusFSO(OmKeyArgs args,
         HeapEntry entry = heapIterator.next();
         OzoneFileStatus status = entry.getStatus(prefixKey,
             scmBlockSize, volumeName, bucketName, replication);
-        map.put(entry.key, status);
```
[the remark this answers is not in the document]
Comment (devmadhuu) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java:217]: Yes @duongkame as per MinHeapIterator implementation.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java [hunk 2]
```diff
@@ -214,11 +215,12 @@ public Collection<OzoneFileStatus> listStatusFSO(OmKeyArgs args,
...
+        map.putIfAbsent(entry.key, status);
       }
     }
 
-    return map.values();
+    return map.values().stream().filter(e -> e != null).collect(
```
[the remark this answers is not in the document]
Comment (devmadhuu) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java:222]: @duongkame  thanks for review. Earlier the null value (means key deleted) from CacheIter was not added in map and was getting lost, so we should allow that and when keyInfo gets null and then getStatus , we'll return null. This was needed to allow deleted key in cache.

**Pull request merged**